x/merkledb/view_iterator.go

```diff
@@ -0,0 +1,170 @@
+// Copyright (C) 2019-2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package merkledb
+
+import (
+	"bytes"
+
+	"github.com/ava-labs/avalanchego/database"
+	"golang.org/x/exp/slices"
+)
+
+func (t *trieView) NewIterator() database.Iterator {
+	return t.NewIteratorWithStartAndPrefix(nil, nil)
+}
+
+func (t *trieView) NewIteratorWithStart(start []byte) database.Iterator {
+	return t.NewIteratorWithStartAndPrefix(start, nil)
+}
+
+func (t *trieView) NewIteratorWithPrefix(prefix []byte) database.Iterator {
+	return t.NewIteratorWithStartAndPrefix(nil, prefix)
+}
+
+func (t *trieView) NewIteratorWithStartAndPrefix(start, prefix []byte) database.Iterator {
+	changes := make([]KeyChange, 0, len(t.changes.values))
+	for path, change := range t.changes.values {
+		key := path.Serialize().Value
+		if (len(start) > 0 && bytes.Compare(start, key) > 0) || !bytes.HasPrefix(key, prefix) {
+			continue
+		}
+		changes = append(changes, KeyChange{Key: key, Value: change.after})
+	}
+	// sort [changes] so they can be merged with the parent trie's state
+	slices.SortFunc(changes, func(a, b KeyChange) bool {
```

Comment:
We could theoretically store these sorted all the time, but I'm not sure if it is worth it.